lib/src/rules/use_iterables.dart

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:analyzer/dart/element/type_provider.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Use lazy iterables.';
+
+const _details = r'''
+**DO** use lazy iterables created by `iterable.map`. Since they are lazily
+instantiated, the `map` function will not be evaluated until use which can be
+confusing.  Alternatively, consider preferring `forEach`.
+
+**BAD:**
+```dart
+[1, 2, 3].map(print);
+```
+
+**GOOD:**
+```dart
+[1, 2, 3].forEach(print);
+[1, 2, 3].map(print).last;
+[1, 2, 3].map(print).toList();
+```
+''';
+
+class UseIterables extends LintRule {
+  UseIterables()
+      : super(
+            name: 'use_iterables',
+            description: _desc,
+            details: _details,
+            group: Group.errors,
+            maturity: Maturity.experimental);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this, context);
+    registry.addMethodInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  final TypeProvider typeProvider;
+
+  _Visitor(this.rule, LinterContext context)
+      : typeProvider = context.typeProvider;
+
+  @override
+  visitMethodInvocation(MethodInvocation node) {
+    if (!_isIterable(node.staticType)) {
+      return;
+    }
+    if (node.methodName.name != 'map') {
+      return;
+    }
+    if (_isUsed(node)) {
+      return;
+    }
+
+    rule.reportLint(node.methodName);
+  }
+
+  bool _isIterable(DartType? type) =>
+      type != null && type.asInstanceOf(typeProvider.iterableElement) != null;
+
+  static bool _isUsed(AstNode node) {
+    var parent = node.parent;
+    if (parent == null) {
+      return false;
+    }
+
+    if (parent is ParenthesizedExpression ||
+        parent is ConditionalExpression ||
+        parent is CascadeExpression) {
+      return _isUsed(parent);
+    }
+
+    return parent is ArgumentList ||
```

Comment:
I'm concerned about both false positives and false negatives. For example, I think the `BinaryExpression` needs to be in this list to prevent a false positive. There are probably others, like `PrefixExpression` and `IndexExpression`. And maybe `ExtensionOverride` needs to be skipped over like `ParenthesizedExpression` above. I'm not sure the list is complete even now.

What we really want is a getter on `Expression` that returns `true` if the value of the expression is used in some way.

Reply:
We definitely want to guard against false positives and negatives so I appreciate the attention to detail!

I added a few kind of odd test cases.  If you can think of some more, please share them and I'll make sure they're covered.

Thanks!

Reply:
Do we have the concept of "being used" around `void` already?
The contexts where we'd allow `void`, which means it isn't being used, are:
* An initializer expression part, or any expression in the increment part, of a `for(;;)` statement.
* The expression of an expression statement.
* Either branch of a conditional expression if the conditional isn't being used.
* A body of parenthesized expression if the parenthesized expression isn't being used.
* The second part of a `??` operator where the `??` expression isn't being used.
* The operand of an `as` expression where the `as` expression isn't isn't being used.
* The operand of an `is` expression.
* The `=> e` expression of a `void`-typed function.
* Assignment to a `void` typed variable (questionable!)
* The one I forgot (there's always one!)

Currently an extension override is safe because you must call a member on it.

(Also the second operand of a `&&` or `||` operator which isn't being used, but types usually prevent that from being relevant).

Also consider `maybeIterable?.map(...)` and `iterable..map(...)` (followed by no selectors).

Reply:
I agree with @lrhn, it would be much better to enumerate the places where the value isn't being used than to try to enumerate the places where it is. Even if it doesn't make the implementation easier it will help prevent false positives when new language constructs are added.

> Do we have the concept of "being used" around `void` already?

I don't believe that we have anything around this that the linter can currently re-use.